File: src/drift/pre-commit.ts
/**
 * Pre-commit hook entry point (R25, R25a). The verbatim hook script
 * (installed by E6) invokes:
 *
 *   claude-profiles drift --pre-commit-warn
 *
 * which lands in this module. Fail-open invariant: this function NEVER
 * throws and always indicates exit code 0. If detection fails for any
 * reason — corrupted state, missing profile dir, FS error mid-walk — we
 * print a one-line diagnostic to stderr and exit silently. A drift check
 * must never block a commit.
 *
 * E5 owns the actual `drift --pre-commit-warn` argv parsing; this module
 * exposes the function E5 dispatches to.
 */

import { detectDrift } from "./detect.js";
import type { StatePaths } from "../state/paths.js";

import type { DriftReport } from "./types.js";

export interface PreCommitWarnResult {
  /** Always 0 — fail-open invariant. */
  exitCode: 0;
  /** Lines that would be written to stderr. Captured for testability. */
  warnings: string[];
  /** The report we produced, or null if detection itself failed. */
  report: DriftReport | null;
}

/**
 * Detect drift; if any, build a one-line + per-file warning. Always returns
 * exit code 0. The caller (E5) wires this into the `drift --pre-commit-warn`
 * code path; it MUST `process.exit(0)` regardless of return value.
 *
 * Output discipline (non-blocking quality bar §7):
 *   - When no drift: silent (no output). The hook should be invisible in
 *     the happy path.
 *   - When drift: a header line ("claude-profiles: <N> drifted file(s) in
 *     .claude/ vs active profile <name>") plus up to 10 entries (truncated
 *     with "...and N more" if longer) so the commit terminal isn't flooded.
 *   - When detection failed: a single line "claude-profiles: drift check
 *     skipped: <reason>". Never two lines, never a stack trace.
 */
export async function preCommitWarn(paths: StatePaths): Promise<PreCommitWarnResult> {
  const warnings: string[] = [];
  let report: DriftReport | null = null;
  try {
    report = await detectDrift(paths);
    // Surface a degraded-state notice (S17 / R42) before drift output.
    // Without it, a corrupted .state.json looks indistinguishable from a
    // fresh project — the user gets no hint that the underlying state
    // file is broken (multi-reviewer second-pass P1). We deliberately
    // skip code "Missing": that's the normal NoActive case for projects
    // that haven't run `init` yet, and we don't want a warning on every
    // commit in those repos.
    if (report.warning && report.warning.code !== "Missing") {
      warnings.push(
        `claude-profiles: state file degraded (${report.warning.code}): ${report.warning.detail}`,
      );
    }
    if (report.fingerprintOk && report.entries.length > 0) {
      warnings.push(...formatWarning(report));
    }
  } catch (err: unknown) {
    const detail = err instanceof Error ? err.message : String(err);
    warnings.push(`claude-profiles: drift check skipped: ${detail}`);
  }

  // EPIPE-safe write loop (multi-reviewer P0-2): if stderr is a closed pipe
  // (e.g. git pre-commit driver disconnected before the hook finishes
  // writing), `process.stderr.write` can throw synchronously. Swallowing
  // the error here preserves the fail-open invariant — better to lose the
  // message than block the commit.
  for (const w of warnings) {
    try {
      process.stderr.write(w + "\n");
    } catch {
      // EPIPE or similar — abandon the rest of the output silently.
      break;
    }
  }
  return { exitCode: 0, warnings, report };
}

const MAX_LINES = 10;

function formatWarning(report: DriftReport): string[] {
  const lines: string[] = [];
  const n = report.entries.length;
  lines.push(
    `claude-profiles: ${n} drifted file(s) in .claude/ vs active profile '${report.active}'`,
  );
  const shown = report.entries.slice(0, MAX_LINES);
  for (const e of shown) {
    lines.push(`  ${statusGlyph(e.status)} ${e.relPath}`);
  }
  if (n > MAX_LINES) {
    lines.push(`  ...and ${n - MAX_LINES} more`);
  }
  return lines;
}

function statusGlyph(
  status: "modified" | "added" | "deleted" | "unrecoverable",
): string {
  switch (status) {
    case "modified":
      return "M";
    case "added":
      return "A";
    case "deleted":
      return "D";
    // cw6/T5: 'X' (broken) for unrecoverable — visually distinct from
    // M/A/D so users immediately see the row needs `init`/`validate`
    // rather than the standard discard/persist gate.
    case "unrecoverable":
      return "X";
  }
}

File: go/internal/drift/pre_commit.go
package drift

import (
	"fmt"
	"io"
	"os"

	"github.com/htxryan/c3p/internal/state"
)

// preCommitMaxLines bounds the per-file lines emitted by PreCommitWarn so a
// commit terminal isn't flooded when a user has dozens of drifted files.
const preCommitMaxLines = 10

// PreCommitWarnResult is the structured outcome of PreCommitWarn. Capturing
// warnings (rather than just writing to stderr) makes the function unit-
// testable. ExitCode is always 0 — fail-open invariant (R25/R25a).
type PreCommitWarnResult struct {
	// ExitCode is always 0 (fail-open). The CLI caller MUST exit 0 regardless
	// of internal errors.
	ExitCode int
	// Warnings are the stderr lines that would be written. Captured for tests.
	Warnings []string
	// Report is the DriftReport we produced, or nil if detection itself failed.
	Report *DriftReport
}

// PreCommitWarn is the pre-commit hook entry point (R25, R25a). The hook
// script (installed by D7's commands) invokes `c3p drift --pre-commit-warn`
// which lands here.
//
// Fail-open invariant: NEVER returns a non-zero exit code. If detection
// fails for any reason — corrupted state, missing profile dir, FS error
// mid-walk — print a one-line diagnostic to stderr and exit silently. A
// drift check must NEVER block a commit.
//
// PR25 (backup notice on non-interactive discard): not relevant here —
// pre-commit never discards; it only warns. The PR25 contract lives in
// ApplyGate's discard path, where SnapshotForDiscard's return path is
// surfaced via ApplyGateResult.BackupSnapshot (*string, nil = no snapshot).
//
// Output discipline (non-blocking quality bar §7):
//   - When no drift: silent (no output). The hook is invisible in the happy
//     path.
//   - When drift: a header line ("c3p: <N> drifted file(s) ...") plus up to
//     10 entries (truncated with "...and N more") so the commit terminal
//     isn't flooded.
//   - When detection failed: a single line "c3p: drift check skipped:
//     <reason>". Never two lines, never a stack trace.
//
// EPIPE-safe write loop: if stderr is a closed pipe (e.g. git pre-commit
// driver disconnected), `os.Stderr.Write` may return an error. Swallow it
// and abandon the rest — preserves the fail-open invariant.
func PreCommitWarn(paths state.StatePaths) PreCommitWarnResult {
	return preCommitWarnTo(paths, os.Stderr)
}

// preCommitWarnTo is the internal entry point that accepts an explicit
// writer. Tests use this directly to assert stderr behavior without the
// data race that a global mutable writer would introduce when t.Parallel
// tests overlap.
func preCommitWarnTo(paths state.StatePaths, w io.Writer) (result PreCommitWarnResult) {
	warnings := []string{}
	var report *DriftReport
	defer func() {
		// Defense in depth: a panic in detection should never block a commit
		// (fail-open). But silent recovery makes future regressions invisible —
		// emit a single diagnostic line so a panicking detector is debuggable.
		// EPIPE: if the writer fails we still return exit 0; the message is
		// best-effort, the fail-open contract is mandatory.
		if r := recover(); r != nil {
			line := fmt.Sprintf("c3p: drift check skipped: panic: %v", r)
			warnings = append(warnings, line)
			_, _ = fmt.Fprintln(w, line)
			result = PreCommitWarnResult{
				ExitCode: 0,
				Warnings: warnings,
				Report:   report,
			}
		}
	}()

	r, err := DetectDrift(paths)
	if err != nil {
		warnings = append(warnings, fmt.Sprintf("c3p: drift check skipped: %v", err))
	} else {
		report = &r
		// Surface a degraded-state notice (S17 / R42) before drift output.
		// Without it, a corrupted .state.json looks indistinguishable from a
		// fresh project. We deliberately skip code "Missing": that's the
		// normal NoActive case for projects that haven't run init yet, and
		// we don't want a warning on every commit in those repos.
		if r.Warning != nil && r.Warning.Code != state.StateReadWarningMissing {
			warnings = append(warnings, fmt.Sprintf(
				"c3p: state file degraded (%s): %s",
				r.Warning.Code, r.Warning.Detail,
			))
		}
		if r.FingerprintOk && len(r.Entries) > 0 {
			warnings = append(warnings, formatPreCommitWarning(r)...)
		}
	}

	// EPIPE-safe write loop: if stderr is a closed pipe, abandon the rest
	// silently — better to lose the message than block the commit.
	for _, line := range warnings {
		if _, werr := fmt.Fprintln(w, line); werr != nil {
			break
		}
	}

	return PreCommitWarnResult{
		ExitCode: 0,
		Warnings: warnings,
		Report:   report,
	}
}

func formatPreCommitWarning(r DriftReport) []string {
	n := len(r.Entries)
	lines := make([]string, 0, n+2)
	active := ""
	if r.Active != nil {
		active = *r.Active
	}
	lines = append(lines, fmt.Sprintf(
		"c3p: %d drifted file(s) in .claude/ vs active profile '%s'",
		n, active,
	))
	limit := n
	if limit > preCommitMaxLines {
		limit = preCommitMaxLines
	}
	for i := 0; i < limit; i++ {
		e := r.Entries[i]
		lines = append(lines, fmt.Sprintf("  %s %s", statusGlyph(e.Status), e.RelPath))
	}
	if n > preCommitMaxLines {
		lines = append(lines, fmt.Sprintf("  ...and %d more", n-preCommitMaxLines))
	}
	return lines
}

func statusGlyph(s DriftStatus) string {
	switch s {
	case DriftStatusModified:
		return "M"
	case DriftStatusAdded:
		return "A"
	case DriftStatusDeleted:
		return "D"
	case DriftStatusUnrecoverable:
		// 'X' (broken) for unrecoverable — visually distinct from M/A/D so
		// users immediately see the row needs init/validate rather than the
		// standard discard/persist gate.
		return "X"
	default:
		return "?"
	}
}
